File: lua/batteries.lua
--[[
  Ring gauges for the batteries card, in the style of the macOS Batteries
  widget: one ring per device, the device's glyph centred inside it, the
  charge under it and the device name under that. Charging devices get a
  bolt badge on the ring, the way macOS badges a charging device green.

  A device that reports two batteries -- the earbuds of any TWS set -- stays a
  SINGLE item and splits its ring instead: the left half of the ring is the
  left bud and the right half is the right bud, each filling from 12 o'clock
  towards 6 and coloured by its own level. Left maps to left, so the shape
  says which is which without a second label.

  Device data comes from bin/batteries.py via a Lua table literal it drops in
  $XDG_RUNTIME_DIR -- see that script for why. This file only draws.

  conky calls conky_batteries_draw() from lua_draw_hook_post, i.e. after the
  card's own text is on screen, so the header drawn by conky and the rings
  drawn here land in the same window.
--]]

require 'cairo'
-- conky >= 1.12 moved cairo_xlib_surface_create into its own module.
pcall(require, 'cairo_xlib')

local HOME = os.getenv('HOME')
local w = dofile(HOME .. '/.config/conky/gruvbox-dark/widget-common.lua')
local D = dofile(HOME .. '/.config/conky/gruvbox-dark/lua/draw.lua')
local CACHE = (os.getenv('XDG_RUNTIME_DIR') or '/tmp')
              .. '/conky-gruvbox/batteries.lua'

-- Must match the layout constants at the top of bin/batteries.py: that script
-- reserves the vertical space these numbers consume.
local COLS    = 3
local RING_D  = 52
local STROKE  = 6
local ROW_GAP = 10
local LABEL_H = 26
local ROW_H   = RING_D + LABEL_H

-- Distance from the top of the window to the top of the first ring. The card's
-- inner margin, plus the height of the header line conky drew above us, plus
-- the HEADER_H gap the script reserved. Measured on this setup at size 10.
local HEAD_LINE = 18
local Y0 = w.PAD + HEAD_LINE + 6

-- Clear space at 12 and 6 o'clock, in radians, that separates the two halves
-- of a split ring so the pair reads as two gauges even at full charge.
local SPLIT_GAP = 0.10
local TOP = -math.pi / 2

local GLYPH_PX = 21
local PCT_PX   = 12
local NAME_PX  = 9

-- Mark for a device whose reading arrived over the network rather than a
-- cable. F05A9 (mdi-wifi) was chosen by RENDERING it in Maple Mono NF CN at
-- NAME_PX, per the warning in bin/batteries.py, and it stays legible there.
-- The check is not ceremony: F0EBE, two codepoints away in the same block,
-- draws a pair of dividers in this font, and F1EB -- the other clean wifi
-- mark -- is a Font Awesome codepoint rather than the Material Design block
-- every other glyph on this card comes from.
local WIFI_GLYPH = '\u{F05A9}'
local WIFI_GAP   = 3

-- gruvbox colour for a charge. Devices that only report a coarse bucket are
-- coloured by the bucket, so a mouse reporting "normal" is not painted amber
-- just because UPower turned that bucket into the number 55.
local function colour(d, pct)
  pct = pct or d.pct
  local p = w.pal
  if d.charging then return p.aqua end
  if d.coarse then
    if d.level == 4 then return p.red end       -- critical
    if d.level == 3 then return p.orange end    -- low
    return p.green                              -- normal / high / full
  end
  if pct > 50 then return p.green end
  if pct > 25 then return p.yellow end
  if pct > 12 then return p.orange end
  return p.red
end

-- One half of a split ring. side = 1 is the right half, sweeping clockwise
-- from 12 o'clock; side = -1 is the left half, sweeping anticlockwise. A nil
-- frac draws only the track, which is how a bud sitting in the case shows up.
local function half_ring(cr, cx, cy, r, side, frac, fill)
  local span = math.pi - 2 * SPLIT_GAP
  local a0 = TOP + side * SPLIT_GAP
  local sweep = function(a1)
    if side < 0 then cairo_arc_negative(cr, cx, cy, r, a0, a1)
    else cairo_arc(cr, cx, cy, r, a0, a1) end
    cairo_stroke(cr)
  end
  D.rgba(cr, w.pal.bg2, 0.6)
  sweep(a0 + side * span)
  if frac and frac > 0.004 then
    D.rgba(cr, fill, 1)
    sweep(a0 + side * span * math.min(frac, 1))
  end
end

-- ── one device ────────────────────────────────────────────────
local function draw_device(cr, d, cx, top, cell)
  local r  = (RING_D - STROKE) / 2
  local cy = top + RING_D / 2
  local p  = w.pal

  cairo_set_line_width(cr, STROKE)
  cairo_set_line_cap(cr, CAIRO_LINE_CAP_ROUND)

  local split = (d.pct_l ~= nil) or (d.pct_r ~= nil)
  if split then
    half_ring(cr, cx, cy, r, -1, d.pct_l and d.pct_l / 100, colour(d, d.pct_l))
    half_ring(cr, cx, cy, r,  1, d.pct_r and d.pct_r / 100, colour(d, d.pct_r))
  else
    D.rgba(cr, p.bg2, 0.6)
    cairo_arc(cr, cx, cy, r, 0, 2 * math.pi)
    cairo_stroke(cr)

    local frac = math.min(math.max(d.pct / 100, 0), 1)
    if frac > 0.001 then
      D.rgba(cr, colour(d), 1)
      -- from 12 o'clock, clockwise, like the macOS ring
      cairo_arc(cr, cx, cy, r, TOP, TOP + 2 * math.pi * frac)
      cairo_stroke(cr)
    end
  end

  D.face(cr, GLYPH_PX, false)
  D.rgba(cr, p.fg1, 0.92)
  D.centred(cr, d.glyph, cx, cy)

  -- Charging badge: a filled disc in the card's own background so the ring
  -- reads as interrupted behind it, then the bolt on top.
  if d.charging then
    local bx, by = cx + r * 0.74, cy + r * 0.74
    D.rgba(cr, p.bg0h, 1)
    cairo_arc(cr, bx, by, 8, 0, 2 * math.pi)
    cairo_fill(cr)
    D.face(cr, 12, false)
    D.rgba(cr, d.full and p.green or p.aqua, 1)
    D.centred(cr, '\u{F140B}', bx, by)   -- lightning-bolt
  end

  local label
  if split then
    -- Both numbers stay visible; they collapse to one when the buds agree,
    -- which they usually do.
    if d.pct_l and d.pct_r and d.pct_l ~= d.pct_r then
      label = d.pct_l .. '/' .. d.pct_r .. '%'
    else
      label = (d.pct_l or d.pct_r) .. '%'
    end
  else
    label = (d.coarse and '~' or '') .. d.pct .. '%'
  end
  D.face(cr, PCT_PX, true)
  D.rgba(cr, p.fg0, 1)
  D.centred(cr, label, cx, top + RING_D + 9)

  -- Name, preceded by the wifi mark when the reading came in over the network.
  -- Without it an unplugged phone is indistinguishable from a ring that has
  -- silently gone stale, which is the one thing this card must not be.
  D.face(cr, NAME_PX, false)
  local name_y = top + RING_D + 21
  if d.network then
    local mark_w = D.extents(cr, WIFI_GLYPH).width
    local name   = D.fit(cr, d.name, cell - 6 - mark_w - WIFI_GAP)
    local e      = D.extents(cr, name)
    -- Baseline is taken from the name alone, so the text sits on exactly the
    -- line it would without the mark and the glyph aligns to it rather than
    -- dragging it around.
    local base   = name_y - e.height / 2 - e.y_bearing
    local x      = cx - (mark_w + WIFI_GAP + e.width) / 2
    D.rgba(cr, p.blue, 1)
    D.left(cr, WIFI_GLYPH, x, base)
    D.rgba(cr, p.fg4, 1)
    D.left(cr, name, x + mark_w + WIFI_GAP, base)
  else
    D.rgba(cr, p.fg4, 1)
    D.centred(cr, D.fit(cr, d.name, cell - 6), cx, name_y)
  end
end

-- ── entry point ───────────────────────────────────────────────
function conky_batteries_draw()
  if conky_window == nil then return end

  local ok, devices = pcall(dofile, CACHE)
  if not ok or type(devices) ~= 'table' or #devices == 0 then return end

  local cs = cairo_xlib_surface_create(conky_window.display, conky_window.drawable,
                                       conky_window.visual,
                                       conky_window.width, conky_window.height)
  local cr = cairo_create(cs)

  -- Fewer than COLS devices spread across the full width rather than huddling
  -- at the left edge, which is what keeps a one- or two-device card balanced.
  local cols = math.min(#devices, COLS)
  local cell = w.WIDTH / cols

  for i, d in ipairs(devices) do
    local col = (i - 1) % cols
    local row = math.floor((i - 1) / cols)
    draw_device(cr, d,
                w.PAD + cell * (col + 0.5),
                Y0 + row * (ROW_H + ROW_GAP),
                cell)
  end

  cairo_destroy(cr)
  cairo_surface_destroy(cs)
  collectgarbage()
end

File: bin/batteries.py
#!/usr/bin/env python3
"""Enumerate every battery UPower knows about, for the conky batteries card.

UPower is the same source xfce4-power-manager reads, so whatever shows up in
its Devices tab shows up here: the laptop pack, HID++ peripherals behind a
Logitech receiver, and Bluetooth devices that publish org.bluez.Battery1.

Two device classes are read elsewhere and merged in, because UPower's own
answer for them is wrong or absent. Both modules are optional -- if either
raises or has nothing current, UPower's view is used unchanged.

  AirPods    what BlueZ publishes is one coarse number, so airpods.py decodes
             Apple's BLE advertisement instead and returns the left bud, right
             bud and case separately; those replace UPower's single entry for
             the same headphones.
  iOS        UPower only sees an iPhone or iPad while it is plugged in, since
             its iDevice backend discovers devices through udev. idevices.py
             goes through usbmux instead, which reaches them over wi-fi too,
             and its entry replaces UPower's whenever both have the device.

Two outputs, both produced in one run:

  * a Lua table literal in $XDG_RUNTIME_DIR that lua/batteries.lua pulls in
    with dofile() to draw the rings. Lua cannot afford to block the draw hook
    on D-Bus round trips, and emitting the consumer's own syntax saves
    shipping a JSON parser written in Lua.
  * conky markup on stdout: the card header, plus a ${voffset} tall enough to
    reserve the ring block. That is what makes the card grow and shrink with
    the number of devices -- conky sizes the window to its text, so reserving
    the space as text is what moves the window edge, and picom rounds
    whatever rectangle results.

Talking to UPower goes through busctl rather than a D-Bus binding so the
script stays stdlib-only, matching bin/weather.py.
"""

import json
import os
import subprocess
import sys

# ── layout, in pixels ─────────────────────────────────────────
# Mirrored by lua/batteries.lua; the two must agree or the rings will not sit
# inside the space reserved for them here. WIDTH matches widget-common.lua.
WIDTH    = 292
COLS     = 3
RING_D   = 52     # ring outer diameter
LABEL_H  = 26     # percentage + device name under each ring
ROW_GAP  = 10
ROW_H    = RING_D + LABEL_H
HEADER_H = 6      # padding under conky's own header line, before the rings
# conky charges a full line height for the line the ${voffset} itself sits on,
# so the reservation has to give that back or the card runs tall. Measured on
# this setup with the default size 9 face.
LINE_H   = 19

BUS   = ["busctl", "--system", "--json=short", "call", "org.freedesktop.UPower"]
IFACE = "org.freedesktop.UPower.Device"
CACHE_DIR = os.environ.get("XDG_RUNTIME_DIR") or f"/run/user/{os.getuid()}"
CACHE = os.path.join(CACHE_DIR, "conky-gruvbox", "batteries.lua")

# UPower DeviceKind -> Nerd Font Material Design glyph.
#
# Every codepoint below was checked by RENDERING it in Maple Mono NF CN, not by
# looking up its name. This font carries an older Material Design Icons block
# than the current upstream table, so plenty of documented codepoints draw
# something unrelated here: F0B3B ("watch") is the digit 2, F0CB8 ("touchpad")
# is a playlist icon, F1041 ("stylus") is a tripod, F17CE ("earbuds") is an
# aerial. Kinds with no glyph that survived that check fall through to
# UNKNOWN_GLYPH rather than being given a wrong picture.
GLYPH = {
    3:  "\U000F0079",  # ups            -> battery
    5:  "\U000F037D",  # mouse
    6:  "\U000F030C",  # keyboard
    8:  "\U000F011C",  # phone          -> cellphone
    9:  "\U000F011C",  # media player
    10: "\U000F04F6",  # tablet
    11: "\U000F0322",  # computer       -> laptop
    12: "\U000F02B4",  # gaming input   -> controller
    13: "\U000F03EB",  # pen            -> pencil
    14: "\U000F037D",  # touchpad       -> mouse (no touchpad glyph renders)
    17: "\U000F02CE",  # headset        (with boom mic, unlike headphones)
    18: "\U000F04C3",  # speakers
    19: "\U000F02CB",  # headphones
    21: "\U000F04C3",  # other audio
    23: "\U000F042A",  # printer
    28: "\U000F00AF",  # bluetooth generic
}
LAPTOP_GLYPH  = "\U000F0322"
UNKNOWN_GLYPH = "\U000F0091"  # battery-unknown

# Ordering: the machine you are sitting at first, then the things you hold,
# then the things you listen through, then whatever else turned up.
KIND_ORDER = {11: 0, 5: 1, 6: 2, 14: 3, 12: 4, 13: 5,
              17: 6, 19: 7, 18: 8, 21: 9, 8: 10, 10: 11}

# Vendor words worth dropping from a model string: they cost horizontal space
# and the glyph already says what the thing is.
VENDORS = ("Logitech", "Sony", "Apple", "Microsoft", "Jabra", "SteelSeries",
           "Bose", "Samsung", "Sennheiser", "Razer", "Corsair", "Dell", "HP",
           "Anker", "Soundcore", "JBL", "Beats", "Google", "Keychron")
# Trailing nouns the glyph already conveys.
SUFFIXES = (" Multi-Device Mouse", " Wireless Mouse", " Mouse", " Keyboard",
            " Headset", " Headphones", " Earbuds", " Speaker", " Bluetooth")


class Unreachable(Exception):
    """UPower could not be talked to at all -- distinct from it reporting no
    batteries. The two must not be confused: treating an outage as an empty
    result would overwrite a good cache and collapse the card."""


def prop_call(path):
    """Properties of one device, or None if that device has just gone away.

    A device vanishing between the enumeration and this call is routine --
    a mouse sleeping, headphones disconnecting -- so it is skipped rather
    than raised on."""
    out = subprocess.run(
        BUS + [path, "org.freedesktop.DBus.Properties", "GetAll", "s", IFACE],
        capture_output=True, text=True, timeout=5)
    if out.returncode != 0:
        return None
    return {k: v["data"] for k, v in json.loads(out.stdout)["data"][0].items()}


def enumerate_paths():
    out = subprocess.run(
        BUS + ["/org/freedesktop/UPower", "org.freedesktop.UPower",
               "EnumerateDevices"],
        capture_output=True, text=True, timeout=5)
    if out.returncode != 0:
        raise Unreachable(out.stderr.strip())
    return json.loads(out.stdout)["data"][0]


def shorten(name):
    name = " ".join(name.split())
    for v in VENDORS:
        if name.lower().startswith(v.lower() + " "):
            name = name[len(v) + 1:]
            break
    for s in SUFFIXES:
        if name.lower().endswith(s.lower()) and len(name) > len(s):
            name = name[:-len(s)]
            break
    # Possessive device names ("James's AirPods Pro") read better without the
    # owner, which is the same for every device here anyway.
    for sep in ("’s ", "'s "):
        if sep in name:
            name = name.split(sep, 1)[1]
            break
    return name.strip() or "Battery"


def collect():
    out = []
    paths = enumerate_paths()
    answered = 0
    for path in paths:
        if path.endswith("/DisplayDevice"):
            continue
        p = prop_call(path)
        if p is None:
            continue          # gone since the enumeration; see prop_call
        answered += 1
        if p.get("Type") == 1 or not p.get("IsPresent"):
            continue          # Type 1 is line power, not a battery

        kind = p.get("Type", 0)
        mains = bool(p.get("PowerSupply"))
        level = p.get("BatteryLevel", 0)
        # BatteryLevel 0 (unknown) / 1 (none) both mean "Percentage is real".
        # Anything else is a coarse bucket and UPower has synthesised a number
        # from it, so the reading gets flagged rather than shown as exact.
        coarse = level not in (0, 1)

        pct = p.get("Percentage")
        if pct is None or (pct <= 0 and coarse):
            continue

        name = shorten(p.get("Model") or p.get("NativePath") or "")
        if mains and kind in (0, 2):
            name, kind = "Laptop", 11   # UPower calls the internal pack "battery"

        out.append({
            "name":     name,
            "glyph":    LAPTOP_GLYPH if mains else GLYPH.get(kind, UNKNOWN_GLYPH),
            "pct":      round(float(pct)),
            "coarse":   coarse,
            "level":    level,
            "charging": p.get("State") in (1, 4, 5),
            "full":     p.get("State") == 4,
            "order":    (0 if mains else 1, KIND_ORDER.get(kind, 99), name),
            "_model":   p.get("Model") or "",
        })

    # Every single device refusing to answer is a bus that dropped out from
    # under us, not a machine that suddenly has no batteries.
    if paths and not answered:
        raise Unreachable("no device answered GetAll")

    out = swap_in_airpods(out)
    out = add_idevices(out)
    out.sort(key=lambda d: d["order"])
    for d in out:
        d.pop("order", None)
        d.pop("_model", None)
    return out


def swap_in_airpods(devices):
    """Replace UPower's single coarse AirPods entry with per-bud and case
    readings, when airpods.py can supply them. Anything going wrong in there
    leaves the UPower entry alone -- it is a refinement, not a dependency."""
    try:
        import airpods
        pods = airpods.read()
    except Exception:
        return devices
    if not pods:
        return devices

    devices = [d for d in devices if d["_model"] != pods["device_name"]]
    # Sorted in with the headphones they replace, and kept in L / R / case order.
    for i, c in enumerate(pods["components"]):
        c["order"] = (1, KIND_ORDER[19], "%d" % i)
        devices.append(c)
    return devices


def add_idevices(devices):
    """Merge in iPhones and iPads reached over usbmux -- which, with netmuxd
    supplying the mDNS discovery stock usbmuxd lacks, means over wi-fi as well
    as USB. See bin/idevices.py.

    UPower's entry for the same device is dropped rather than kept alongside:
    it appears only while the device is plugged in, so keeping both would make
    the phone show up twice at exactly the moment it is cabled."""
    try:
        import idevices
        found = idevices.read()
    except Exception:
        return devices
    if not found:
        return devices

    def key(name):
        return " ".join((name or "").split())

    # .get, not [], because swap_in_airpods has already run and its per-bud
    # rows are synthesised, not UPower's -- they carry no _model at all.
    seen = {key(d["_model"]) for d in found}
    devices = [d for d in devices if key(d.get("_model")) not in seen]
    for d in found:
        d["order"] = (1, KIND_ORDER.get(d.pop("_kind"), 99), d["name"])
        devices.append(d)
    return devices


def lua_str(s):
    """Quote a Python str as a Lua string. Lua strings are byte strings and the
    cache file is read as UTF-8, so the glyphs can go in literally."""
    out = s.replace("\\", "\\\\").replace('"', '\\"')
    return '"' + "".join(c if c >= " " else "\\%d" % ord(c) for c in out) + '"'


def lua_value(v):
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return repr(v)
    return lua_str(v)


def write_cache(devices):
    os.makedirs(os.path.dirname(CACHE), exist_ok=True)
    rows = ",\n".join(
        "  {" + ", ".join(f"{k}={lua_value(v)}"
                          for k, v in d.items() if v is not None) + "}"
        for d in devices)
    tmp = CACHE + ".tmp"
    with open(tmp, "w") as fh:
        fh.write("return {\n" + rows + ("\n" if rows else "") + "}\n")
    os.replace(tmp, CACHE)


def show(devices):
    for d in devices:
        flag = "~" if d["coarse"] else " "
        chg = " (charging)" if d["charging"] else ""
        print(f"  {d['glyph']}  {d['name']:<18} {flag}{d['pct']:>3}%{chg}")


def main():
    if "--show" in sys.argv:
        show(collect())
        return 0

    try:
        devices = collect()
    except Exception:
        # A card that silently keeps its last good drawing beats one that
        # prints a traceback into the desktop.
        devices = None

    if devices is not None:
        try:
            write_cache(devices)
        except OSError:
            pass
        n = len(devices)
    else:
        # Enumeration failed; keep the last good cache and reserve the space it
        # still needs, so the card holds its size instead of collapsing.
        try:
            with open(CACHE) as fh:
                n = sum(1 for line in fh if line.startswith("  {"))
        except OSError:
            n = 0

    rows = max(1, -(-n // COLS))                      # ceil
    body = HEADER_H + rows * ROW_H + (rows - 1) * ROW_GAP if n else HEADER_H
    body = max(0, body - LINE_H)

    print("${color1}${font Maple Mono NF CN:size=10}\U000F0079${font}  "
          "${color5}${font Maple Mono NF CN:size=8}BATTERIES${font}")
    if not n:
        print("${voffset 2}${color5}no batteries reported")
    print("${voffset %d}" % body, end="")
    return 0


if __name__ == "__main__":
    sys.exit(main())
